source/common/common/utility.h

```diff
@@ -636,15 +636,100 @@ class WelfordStandardDeviation {
   double m2_{0};
 };
 
-template <class Value> struct TrieEntry {
+// BigTrieEntry is used for very simple very fast lookups.
+//
+// Each node contains 256 pointers to potential child nodes, making an
+// 8KB allocation for every node. Since the vast majority of nodes are
+// very sparse, this is extremely space-inefficient, but it makes for
+// the fastest possible lookups if the total table size is small.
+// At larger table sizes, memory cache misses come into play and a more
+// compact trie actually runs faster.
+//
+// Benchmarks suggest that for the important use case of
+// the static header lookup table, BigTrieEntry lookups run in about
+// 2/3 the time that SmallTrieEntry lookups do, but for cases like
+// large route tables, or with long keys, BigTrieEntry runs in around
+// 2x the time and 100x the memory.
+//
+// BigTrieEntry should therefore should not be used in contexts where
+// performance is not critical, or where the contents may be large.
+template <class Value> class BigTrieEntry {
+public:
+  BigTrieEntry* operator[](uint8_t i) const { return entries_[i].get(); }
+  void set(uint8_t branch, std::unique_ptr<BigTrieEntry<Value>> entry) {
+    entries_[branch] = std::move(entry);
+  }
   Value value_{};
-  std::array<std::unique_ptr<TrieEntry>, 256> entries_;
+
+private:
+  std::array<std::unique_ptr<BigTrieEntry>, 256> entries_;
+};
+
+// A SmallTrieEntry aims to be a good balance of performant and
+// space-efficient, by allocating a vector the size of the range of children
+// the node contains. This should be good for most use-cases.
+//
+// For example, a node with children 'a' and 'z' will contain a vector of
+// size 26, containing two values and 24 nulls. A node with only one
+// child will contain a vector of size 1. A node with no children will
+// contain an empty vector.
+//
+// Compared to allocating 256 entries for every node, this makes insertions
+// a little bit inefficient (especially insertions in reverse order), but
+// trie lookups remain O(length-of-longest-matching-prefix) with just a
+// couple of very cheap operations extra per step.
+//
+// By size, having 256 entries for every node makes each node's overhead
+// (excluding values) consume 8KB; even a trie containing only a single
+// prefix "foobar" consumes 56KB.
+// Using ranged vectors like this makes a single prefix "foobar" consume
+// less than 20 bytes per node, for a total of 0.14KB.
+template <class Value> class SmallTrieEntry {
+public:
+  Value value_{};
+  SmallTrieEntry* operator[](uint8_t i) const {
```

Comment:
Maybe have 2 methods: a const method that returns a const ptr, and a non-const method that returns a non-const ptr?

Reply:
I don't think that even does anything, if both functions have the same behavior? A client that's asking for a const will get the exact same thing from a non-const-returning function as it would if both functions existed.

And it's effectively a private operator, in that the Trie class never exposes a TrieEntry.

If we go with my suggestion of fixing the static header map first then it can be made more explicit that this is private, in that the entry template could be a private member class of the trie template.

Reply:
It's essentially to ensure const-correctness (see https://chromium.googlesource.com/chromium/src/+/main/styleguide/c++/const.md for more details).

Reply:
Ah, that makes more sense, I thought the suggestion was const and non-const just on the return value, `const X fn() const;` and `X fn();` is much more coherent. Will do.